### PR TITLE
Added ability to set alternate keys in fluent syntax

### PR DESCRIPTION
**Pull request by nsarris**
Fix for issue #421 

Added an alternative to Key() named AltKey() that can be used with both an anonymous object or a dictionary (names are mandatory)

Followed the original pattern where non matching keys are ignored

Added an associated test

### CODE REVIEW

File: src/Simple.OData.Client.Core/Fluent/IFluentClient.cs [hunk 1]
```diff
@@ -71,6 +71,18 @@ public interface IFluentClient<T, FT>
         /// <returns>Self.</returns>
         FT Key(T entryKey);
 
+        /// Sets the specified alternate entry key.
```
Comment (watfordgnf) [inline src/Simple.OData.Client.Core/Fluent/IFluentClient.cs]: Missing a `<summary>` tag.

File: src/Simple.OData.Client.Core/Fluent/IFluentClient.cs [hunk 2]
```diff
@@ -71,6 +71,18 @@ public interface IFluentClient<T, FT>
...
+        /// </summary>
+        /// <param name="entryKey">The entry key.</param>
+        /// <returns>Self.</returns>
+        FT AltKey(object entryKey);
+        /// <summary>
```
Comment (watfordgnf) [inline src/Simple.OData.Client.Core/Fluent/IFluentClient.cs]:
```suggestion

        /// <summary>
```

File: src/Simple.OData.Client.Core/Fluent/IFluentClient.cs [hunk 3]
```diff
@@ -71,6 +71,18 @@ public interface IFluentClient<T, FT>
         /// <returns>Self.</returns>
         FT Key(T entryKey);
 
+        /// Sets the specified alternate entry key.
+        /// </summary>
+        /// <param name="entryKey">The entry key.</param>
```
Comment (watfordgnf) [inline src/Simple.OData.Client.Core/Fluent/IFluentClient.cs]:
```suggestion
        /// <param name="entryKey">The alternate entry key.</param>
```

File: src/Simple.OData.Client.Core/Fluent/IFluentClient.cs [hunk 4]
```diff
@@ -71,6 +71,18 @@ public interface IFluentClient<T, FT>
...
+        /// <returns>Self.</returns>
+        FT AltKey(object entryKey);
+        /// <summary>
+        /// Sets the specified alternate entry key.
+        /// </summary>
+        /// <param name="entryKey">The entry key.</param>
```
Comment (watfordgnf) [inline src/Simple.OData.Client.Core/Fluent/IFluentClient.cs]:
```suggestion
        /// <param name="entryKey">The alternate entry key.</param>
```

File: src/Simple.OData.Client.Core/Fluent/FluentCommand.cs [hunk 5]
```diff
@@ -258,6 +260,31 @@ public FluentCommand Key(IDictionary<string, object> key)
...
+        }
+
+        public FluentCommand AltKey(object key)
+        {
+            if (IsBatchResponse) return this;
+
+            if (key != null && TypeCache.IsAnonymousType(key.GetType()))
+            {
+                _details.NamedKeyValues = TypeCache.ToDictionary(key);
+            }
```
Comment (watfordgnf) [inline src/Simple.OData.Client.Core/Fluent/FluentCommand.cs]: This should probably raise an argument exception if an object is used that cannot be made into a `NamedKeyValues` dictionary. As this stands it would also not support `client.AltKey("name")` looking up that there is only a single alternate key; an exception should be raised in that instance too.

File: src/Simple.OData.Client.Core/Fluent/FluentCommand.cs [hunk 6]
```diff
@@ -579,30 +606,25 @@ private string Format()
                     return null;
 
                 var keyNames = _details.Session.Metadata.GetDeclaredKeyPropertyNames(this.EntityCollection.Name).ToList();
-                var namedKeyValues = new Dictionary<string, object>();
-                for (var index = 0; index < keyNames.Count; index++)
+
+                return (_details.NamedKeyValues?.Select(x => new
```
Comment (watfordgnf) [inline src/Simple.OData.Client.Core/Fluent/FluentCommand.cs]: Out of curiosity, why was this changed from the original implementation?

File: src/Simple.OData.Client.UnitTests/FluentApi/FindTypedTests.cs [hunk 7]
```diff
@@ -287,6 +287,17 @@ public async Task Get()
             Assert.Equal(1, category.CategoryID);
         }
 
+        [Fact]
+        public async Task GetWithAltKey()
```
Comment (watfordgnf) [inline src/Simple.OData.Client.UnitTests/FluentApi/FindTypedTests.cs]: A test which uses the `IDictionary<string, object>` overload should be included as well.

File: src/Simple.OData.Client.Core/Fluent/CommandDetails.cs [hunk 8]
```diff
@@ -14,6 +14,7 @@ class CommandDetails
         public string DynamicPropertiesContainerName { get; set; }
         public string FunctionName { get; set; }
         public string ActionName { get; set; }
+        public bool IsAlternateKey { get; set; }
```
Comment (watfordgnf) [inline src/Simple.OData.Client.Core/Fluent/CommandDetails.cs]: A comment pointing to the current community spec would be helpful for future editors.

File: src/Simple.OData.Client.Core/Fluent/IFluentClient.cs (hunk 1 above)
Comment (nsarris) [reply src/Simple.OData.Client.Core/Fluent/IFluentClient.cs]: Fixed

File: src/Simple.OData.Client.Core/Fluent/IFluentClient.cs (hunk 2 above)
Comment (nsarris) [reply src/Simple.OData.Client.Core/Fluent/IFluentClient.cs]: I noticed that overloads are grouped together without spaces so I kept it like that. Let me know if you still want that change

File: src/Simple.OData.Client.Core/Fluent/IFluentClient.cs (hunk 3 above)
Comment (nsarris) [reply src/Simple.OData.Client.Core/Fluent/IFluentClient.cs]: Fixed

File: src/Simple.OData.Client.Core/Fluent/FluentCommand.cs (hunk 5 above)
Comment (nsarris) [reply src/Simple.OData.Client.Core/Fluent/FluentCommand.cs]: That was my initial thought but reading through the code (for that existing Key() implementation) I noticed that
- No validation is performed when setting the values using Key()
- Validation is only performed when building the CommandText at which point the only thing done is filtering out invalid names (if named pairs given) or zipping together the values (if values given)
- Alternate Keys are not handled in the metadata

So I followed that pattern. If you want I can perform another PR that handles Alternate Key meta and performs actual validation on both Primary or Alternate keys.  For alternate keys I could also make the discovery of FilterAsKey working for alternate keys.

File: src/Simple.OData.Client.Core/Fluent/FluentCommand.cs (hunk 6 above)
Comment (nsarris) [reply src/Simple.OData.Client.Core/Fluent/FluentCommand.cs]: The original implementation was:
- When named values given, filter out invalid names using the primary key definition
- When values given, zip together with the primary key names

I tried to perform minimal changes so I tried to squeeze alternate keys in there too. The intent was to discover matching model fields so that I only keep the original behaviour for primary and not alternate keys. Looking at it again I forgot that filter so I commited a fix and relevant tests for that, replicating the original behaviour (filter out invalid names when primary key). If we perform the validation discussed on the previous review item then this can go.

File: src/Simple.OData.Client.Core/Fluent/IFluentClient.cs (hunk 4 above)
Comment (nsarris) [reply src/Simple.OData.Client.Core/Fluent/IFluentClient.cs]: Fixed

File: src/Simple.OData.Client.UnitTests/FluentApi/FindTypedTests.cs (hunk 7 above)
Comment (nsarris) [reply src/Simple.OData.Client.UnitTests/FluentApi/FindTypedTests.cs]: I refactored the object overload to call the dictionary overload underneath so the test covers both cases. Let me know if still want a seperate test.

File: src/Simple.OData.Client.Core/Fluent/CommandDetails.cs (hunk 8 above)
Comment (nsarris) [reply src/Simple.OData.Client.Core/Fluent/CommandDetails.cs]: Please elaborate, I'm not sure what I have to do here :)

File: src/Simple.OData.Client.Core/Fluent/IFluentClient.cs (hunk 2 above)
Comment (watfordgnf) [reply src/Simple.OData.Client.Core/Fluent/IFluentClient.cs]: No that's okay if that's the original way.

File: src/Simple.OData.Client.Core/Fluent/FluentCommand.cs (hunk 5 above)
Comment (watfordgnf) [reply src/Simple.OData.Client.Core/Fluent/FluentCommand.cs]: If there is not any validation along the `Key` path, I'm okay with this as-is for this PR. Certainly a subsequent PR could include the validation.

File: src/Simple.OData.Client.Core/Fluent/CommandDetails.cs (hunk 8 above)
Comment (watfordgnf) [reply src/Simple.OData.Client.Core/Fluent/CommandDetails.cs]: I was unfamiliar with the alternate keys syntax. Basically adding a comment with a pointer to the spec that governs them would be helpful: https://github.com/OData/vocabularies/blob/master/OData.Community.Keys.V1.md

File: src/Simple.OData.Client.Core/Fluent/FluentCommand.cs (hunk 5 above)
Comment (nsarris) [reply src/Simple.OData.Client.Core/Fluent/FluentCommand.cs]: In that case, now that I think about it a separate AltKey method would be obsolete. OData it self has one syntax for primary and alternate key so let's follow that. Let me follow up on this PR and consume the alternate key metadata to have the Key method handle everything.

File: src/Simple.OData.Client.Core/Fluent/CommandDetails.cs (hunk 8 above)
Comment (nsarris) [reply src/Simple.OData.Client.Core/Fluent/CommandDetails.cs]: Alright, want that comment right there on the IsAlternateKey property?

**Comment by nsarris** [comment]
Refactored the code to use the Key() syntax (removed the AltKey syntax altogether). This mimics the behaviour of the odata spec on keys and alternate keys, while being backwards compatible with Simple.

- A collection of values is treated as a primary key by zipping the values together with the names. If less values are provided the query will fail on execution. If more are provided the extra values are ignored.
- A key/value pair collection scans all keys (primary first) for exact matches and selects the first occurance. If no match is found it falls back to trying to extract the primary key from the key/value pairs. This is necessary for backwards compatibility (this is how a whole entity passed in Key is treated).

We can handle the entity instance as key seperately (by making sure the type or type name match before treating it as a key). Then it would be possible to actually throw an exception when no other matches are found. I suggest we talk about this in another PR.

I added a collection of tests for any combination of keys I could think about.

File: src/Simple.OData.Client.Core/Fluent/CommandDetails.cs (hunk 8 above)
Comment (watfordgnf) [reply src/Simple.OData.Client.Core/Fluent/CommandDetails.cs]: After reviewing your updated changes I think this would be best in a the V4 Metadata code by the `GetAlternateKeyPropertyNames`.

**Comment by watfordgnf** [comment]
Agreed on using another PR to tackle some of the validation logic (one I considered was calling `Key` without any primary key in the case where no alternate keys exist).

File: src/Simple.OData.Client.Core/Fluent/CommandDetails.cs (hunk 8 above)
Comment (nsarris) [reply src/Simple.OData.Client.Core/Fluent/CommandDetails.cs]: Added summary comments to interface declaration and both V3/V4 implementations of GetAlternateKeyPropertyNames. It would be nice if this were done for all the metadata methods.

File: src/Simple.OData.Client.V4.Adapter/Metadata.cs [hunk 9]
```diff
@@ -229,10 +229,24 @@ public override IEnumerable<string> GetDeclaredKeyPropertyNames(string collectio
...
+        public override IEnumerable<IEnumerable<string>> GetAlternateKeyPropertyNames(string collectionName)
+        {
+            var entityType = GetEntityType(collectionName);
+            return _model.GetAlternateKeysAnnotation(entityType)
+                .Select(x => x.Select(y => y.Key));
         }
 
         public override string GetFunctionFullName(string functionName)
         {
+            GetAlternateKeyPropertyNames("");
```
Comment (watfordgnf) [inline src/Simple.OData.Client.V4.Adapter/Metadata.cs]: Was this testing code, or is there a side effect here that is important?

Comment (nsarris) [reply src/Simple.OData.Client.V4.Adapter/Metadata.cs]: No idea how that got there, possible an accidental paste. Removed

**Comment by object** [comment]
Thanks for the PR. Tested and merged.
